server/CGameHandler.cpp

```diff
@@ -3920,15 +3920,15 @@ bool CGameHandler::sacrificeCreatures(const IMarket * market, const CGHeroInstan
 	return true;
 }
 
-bool CGameHandler::sacrificeArtifact(const IMarket * m, const CGHeroInstance * hero, const std::vector<ArtifactInstanceID> & arts)
+bool CGameHandler::sacrificeArtifact(const IMarket * market, const CGHeroInstance * hero, const ObjectInstanceID & marketId, const std::vector<ArtifactInstanceID> & arts)
```

Comment:
Why both pointer to IMarket and ObjectInstanceID are needed? It should be possible to obtain either one from another

Reply:
I thought that passing ids directly is better than a certain number of dynamic_casts. Perhaps even IMarket should be ObjectInstance. Because IMarket is always an object on the map.

Reply:
Perhaps instead add method to CGObjectInstance or to IObjectInterface directly:
```cpp
const IMarket * CGObjectInstance::getMarket() const
{
    return dynamic_cast<const IMarket*>(this);
}
```
Market and shipyard already act as optional extensions to IObjectInterface, so makes sense to integrate them better. And I would like to have few more such extensions in future.

Alternatively, if we want to minimize overhead of dynamic_cast we can turn it into virtual method - CGObjectInstance will always return nullptr, and CMarketInstance (or what's its name) will return `this`, removing need of dynamic cast

Reply:
> so makes sense to integrate them better

Do you mean composition?

Reply:
Not necessarily.
If we go virtual function route, then `CGMarkeInstance::getMarket()` can be implemented as either `return this` in case of inheritance or `return market` in case of composition. This would be implementation detail of CGMarketInstance.

Outside code would always use CGObjectInstance::getMarket(). Ideally - outside code should not even access CGMarketInstance, and always use getMarket() method.

Reply:
In this case, we will still need to additionally pass the object id for methods that require a market.

Reply:
We'll need to pass *either* object ID or CGObjectInstance pointer. Not both. `getMarket` would be a virtual method of IObjectInterface.

If we go for object ID then you can retrieve market via something like `cb->getObject(objectID)->getMarket()`, if you have CGObjectInstance - then simply via `object->getMarket()`

Reply:
If we inherit CMarketInstance from IObjectInterface, and then CGTownInstance inherits from CMarketInstance. CGTownInstance must virtually inherit from IObjectInterface. Is it ok?

Reply:
> Or, alternatively, we could also add `getObject` method to IMarket that would return either CGObjectInstance or IObjectInterface.

But the market does not own information about CGObjectInstance or IObjectInterface. Should the market inherit from CGObjectInstance? Or include it?

Reply:
`getObject` method will have to be virtual and implemented by CGMarket / CGTownInstance as simple `return this`

Reply:
This solution is cleaner than the previous one. I like it.